gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/FlowEdge.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flowgraph;
+
+import java.util.List;
+
+import org.apache.gobblin.runtime.api.FlowTemplate;
+import org.apache.gobblin.runtime.api.SpecExecutor;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import com.typesafe.config.Config;
+
+/**
+ * Representation of an edge in a FlowGraph. Each {@link FlowEdge} encapsulates:
+ * <p><ul>
+ *   <li> two {@link DataNode}s as its end points
+ *   <li>a {@FlowTemplate} that responsible for data movement between the {@DataNode}s.
+ *   <li> a list of {@link SpecExecutor}s where the {@link FlowTemplate} can be executed.
+ * </ul></p> and
+ *
+ */
+public interface FlowEdge {
+  /**
+   *
+   * @return the {@link DataNode} ids that are the end points of the edge.
+   */
+  List<String> getEndPoints();
+
+  /**
+   *
+   * @return the {@link FlowTemplate} that performs the data movement along the edge.
+   */
+  FlowTemplate getFlowTemplate();
+
+  /**
+   *
+   * @return a list of {@link SpecExecutor}s that can execute the {@link FlowTemplate} corresponding to this edge.
+   */
+  List<SpecExecutor> getExecutors();
```

Comment:
Similar reasoning as in `DataNode`, I would suggest move this out of `FlowEdge`. The association could be moved to `FlowGaph`.

Reply:
Actually, SpecExecutors are associated with FlowEdge i.e. it represents executor instance where the flow corresponding to the edge can be executed. Moving it to FlowGraph will need yet another map to maintain association between FlowEdge and SpecExecutor, which needs to be updated with every graph mutation.